### PR TITLE
Internal DDS improvements

### PR DESCRIPTION
**Pull request by jrw972**
Next round of Internal DDS improvements.

### CODE REVIEW

File: dds/DCPS/InternalDataReader.h [hunk 1]
```diff
@@ -28,31 +30,65 @@ namespace DCPS {
...
+    x.disposed_generation_count == y.disposed_generation_count &&
+    x.no_writers_generation_count == y.no_writers_generation_count &&
+    x.sample_rank == y.sample_rank &&
+    x.generation_rank == y.generation_rank &&
+    x.absolute_generation_rank == y.absolute_generation_rank &&
+    x.valid_data == y.valid_data;
+}
+#endif
+
+class SIW {
```
Comment (iguessthislldo) [inline dds/DCPS/InternalDataReader.h:79]: This seems a bit too terse. Also I don't know how extensive this header is, but a short all caps identifier like this seems vulnerable to be messed by the predecessor. What about `SIWrap`?

File: dds/DCPS/InternalDataReader.h [hunk 2]
```diff
@@ -184,62 +192,257 @@ class InternalDataReader : public InternalEntity {
...
+        ++no_writers_generation_count_;
+        break;
+      }
+
+      instance_state_ = DDS::ALIVE_INSTANCE_STATE;
 
-        if (publication_set.empty() && instance_state != DDS::NOT_ALIVE_DISPOSED_INSTANCE_STATE) {
-          instance_state = DDS::NOT_ALIVE_NO_WRITERS_INSTANCE_STATE;
+      if (qos.history.kind == DDS::KEEP_LAST_HISTORY_QOS) {
+        while (read_samples_.size() + not_read_samples_.size() >= qos.history.depth) {
```
Comment (iguessthislldo) [inline dds/DCPS/InternalDataReader.h:355]: More warnings:



comparison of integer expressions of different signedness: ‘std::__cxx11::list<OpenDDS::DCPS::InternalDataReader<OpenDDS::DCPS::NetworkInterfaceAddress>::Instance::SampleHolder, std::allocator<OpenDDS::DCPS::InternalDataReader<OpenDDS::DCPS::NetworkInterfaceAddress>::Instance::SampleHolder> >::size_type’ {aka ‘long unsigned int’} and ‘const Long’ {aka ‘const int’} [-Wsign-compare]

File: dds/DCPS/InternalDataReader.h [hunk 3]
```diff
@@ -184,62 +192,257 @@ class InternalDataReader : public InternalEntity {
...
+
+      for (typename SampleList::const_iterator pos = read_samples_.begin(), limit = read_samples_.end();
+           pos != limit; ++pos) {
+        samples.push_back(pos->sample);
+        infos.push_back(make_sample_info(DDS::READ_SAMPLE_STATE, view_state_, instance_state_, pos->disposed_generation_count, pos->no_writers_generation_count, 0, 0, 0, pos->valid_data));
+        ++sample_count;
+      }
+      read_samples_.clear();
+
+      for (typename SampleList::const_iterator pos = not_read_samples_.begin(), limit = not_read_samples_.end();
```
Comment (iguessthislldo) [inline dds/DCPS/InternalDataReader.h:318]: Maybe this doesn't matter or I'm not getting something, but `read` goes over `not_read_samples_`, then `read_samples_`. Why is `take` is doing it in the opposite order, `read_samples_`, then `not_read_samples_`?

**Review by iguessthislldo — CHANGES_REQUESTED** [review]
This is a partial review, I want to do another pass after I've had sleep.

File: dds/DCPS/InternalDataReader.h (hunk 3 above)
Comment (jrw972) [reply dds/DCPS/InternalDataReader.h:318]: When reading, the goal is to get samples to the read state.
When taking, the goal is to remove samples from the data reader so we start with ones that have already been read.
Ultimately, these methods will take a count and masks that limits the number of samples so the difference will become more relevant.

File: dds/DCPS/Qos_Helper.h [hunk 4]
```diff
@@ -1053,6 +1021,106 @@ class WriterDataLifecycleQosPolicyBuilder {
...
 };
 
+class ReaderDataLifecycleQosPolicyBuilder {
+public:
+  ReaderDataLifecycleQosPolicyBuilder()
+  {
+    qos_.autopurge_nowriter_samples_delay.sec = DDS::DURATION_INFINITE_SEC;
+    qos_.autopurge_nowriter_samples_delay.nanosec = DDS::DURATION_INFINITE_NSEC;
+    qos_.autopurge_disposed_samples_delay.sec = DDS::DURATION_INFINITE_SEC;
+    qos_.autopurge_disposed_samples_delay.nanosec = DDS::DURATION_INFINITE_NSEC;
```
[the remark this answers is not in the document]
Comment (jrw972) [reply dds/DCPS/Qos_Helper.h:1031]: Builder support in the idl compiler sounds interesting.  The immediate concern that comes to mind is how to handle nesting.

File: tests/unit-tests/dds/DCPS/InternalDataWriter.cpp [hunk 5]
```diff
@@ -1,8 +1,11 @@
 #include <dds/DCPS/InternalDataWriter.h>
 
+#include <dds/DCPS/Qos_Helper.h>
+
 #include <gtest/gtest.h>
 
 using namespace OpenDDS::DCPS;
+using SIW = SampleInfoWrapper;
```
Comment (iguessthislldo) [inline tests/unit-tests/dds/DCPS/InternalDataWriter.cpp:8]:
```suggestion
typedef SampleInfoWrapper SIW;
```

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

**Review by iguessthislldo — APPROVED** [review]

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

**Review by iguessthislldo — APPROVED** [review]

**Pull request merged**